### PR TITLE
Release names and release criteria

### LINKED ISSUE #13: Release/Versioning
What do we want to label the fall release?

For a client CLI, the version is less important as you can state the user must upgrade to a newer CLI version.

However, registries will store these manifests for an indefinite amount of time. 

- We assume there will be newer additions,
- We want a versioning schema that allows newer clients to differentiate between versions

Questions:
- Do we want to call this a `0.1` release?
- Do we call this 1.0 release, and any non-breaking changes would be 1.1, while larger, possibly breaking changes that registries should opt-into would be 2.0?

Where this matters:

The artifact manifest has a `mediaType` of `application/vnd.cncf.oras.artifact.manifest.v1+json`
What should the value of .v**N**+json be?

### PR DESCRIPTION
**Pull request by SteveLasker**
Adds names of releases, and sets a triage bar for what can be changed within a draft, a release candidate, vs. a major and minor release.

Resolves #13 

Signed-off-by: Steve Lasker <stevenlasker@hotmail.com>

### CODE REVIEW

**Comment by sajayantony** [comment]
Looking at other CNCF projects, listed below - I see for releasing binaries and images/artifacts `alpha` makes sense to me. Draft of a specification is what I can relate with in my earlier life. 

https://github.com/istio/enhancements/issues/12
https://registry.terraform.io/providers/hashicorp/kubernetes-alpha/latest
https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/README.md
https://www.cncf.io/blog/2017/04/24/release-update-prometheus-1-6-1-sneak-peak-2-0/

/cc @lachie83 for inputs on this since we want to start releasing oras and a version of distribution that supports reference-types.

File: release-criteria.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+1.n releases represent minor additive changes to the 1.0 spec. By providing 1.n versioning, the client and registry clearly know there may be new content or capabilities included. By providing 1.n versioning, the artifacts-spec enables reliable expectations between the various registry clients and the registry services.
+
+### 1.n - Acceptable Issues and PRs
+
+1. Additional properties that standardize communications in transit. 
+2. Clarifications on properties, schemas and apis that fill gaps in the expectations of usage. Clarifications SHOULD NOT block scenarios that are declared obvious usages of a previous minor release.
+
+### 0.n - Non-acceptable PRs
+
+1. Additional properties that may leak expectations a server may be processing the property.
```
Comment (sajayantony) [inline release-criteria.md:62]: This is mixing release criteria with Governance and needs to be split.

Comment (SteveLasker) [reply release-criteria.md:62]: Can you elaborate on why this is governance? 
This is setting a bar for what types of changes would be accepted, as a bar. The maintainers can agree, through super majority to just about anything. So, is this in conflict, or simply providing structure to work within?

Comment (jonjohnsonjr) [reply release-criteria.md:62]: I'm having a hard time understanding what this sentence means, myself.

Comment (sajayantony) [reply release-criteria.md:62]: I'm trying to understand what is the scope of the PR and why we are defining a PR acceptance criteria in the release guidance. 
Pull request guidelines should be a contributing guideline and scope is open to discussion as long as it is as per CNCF code of conduct. For e.g. - https://github.com/helm/helm/blob/main/CONTRIBUTING.md

Comment (sajayantony) [reply release-criteria.md:62]: @lachie83  - pointed me to this one as well - https://github.com/cnabio/cnab-spec/blob/main/901-process.md

File: release-criteria.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+The `1.0-draft[n]` releases will be used for end to end (e2e) validations, prior to a level of confidence content may be indefinitely persisted. 
+
+The `-draft[n]` releases are intended to lock-down potential future breaking changes, close on schemas, APIs and parameter names.
+
+Content persisted with a `-draft[n]` version has no expectation it may be supported in a future release.
+
+Content that is persisted (pushed) with `-draft[n]` MAY fail PULL as the registry completes the release. Users must always be able to delete persisted content. However, depending on the specific capability, this may require deleting the repo within the registry.
+
+Content persisted with these versions MUST use the `"mediaType": application/vnd.cncf.oras.artifact.manifest.1.0-draft[n].json` version to assure it's not assumed to be compliant with committed, released builds.
```
Comment (michaelb990) [inline release-criteria.md:17]: nit: you're mixing `"mediaType"` in quotes & `application/vnd.cncf.oras.artifact.manifest.1.0-draft[n].json` not quoted here. 

For OCI media types, we simply use `v1`. My understanding of this is that all you *should* have to know is the major version of the manifest in order to know whether or not you can parse it, as ANY additional changes to v1 manifests MUST be non-breaking. Will we make similar guarantees in ORAS? I think we should.

Also, the [standard way](https://en.wikipedia.org/wiki/Media_type) I've seen this written is to have the type/subtype of the media type and then a format suffix separated by a `+`. I'd propose changing the mediaType to:

`application/vnd.cncf.oras.artifact.manifest.v1-draft[n]+json`

File: release-criteria.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+### 1.0-draft[n] - Acceptable issues and PRs
+
+1. Additional properties
+1. Removing Properties
+1. Property Renames
+1. Breaking changes to persisted content
+1. Additional clarification that may loosen definitions of what may be persisted
+1. Additional clarification that may restrict content previously enabled
+1. Samples and docs that clarify use cases
+1. Issues related to content copying within and across registry implementations
```
Comment (michaelb990) [inline release-criteria.md:28]: This one seems pretty specific, can we loosen it to be more generic? What about issues related to the content lifecycle? Annotations? Descriptors? 

Actually, what if we just removed this? Why do we need to say what is and isn't an acceptable issue? Is that something that we're following from other projects? From my POV, an issue is always acceptable, but whether we will merge a PR for it now or later is what we need guidance for.

File: release-criteria.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+A target of 2-3 `-RC[n]` releases will be made, enabling stabilization with additional, non-breaking changes to persisted content and APIs for storing, discovering or retrieving content.
+
+### rc[n] - Acceptable Issues and PRs
+1. Additional properties
+1. Additional clarification that may loosen definitions of what may be persisted
+1. Samples and docs that clarify use cases
+
+### rc[n] - Non-acceptable PRs
+1. API, Parameter, Existing Schema Renames
+2. Changes that may cause failures in retrieving artifact-manifest content, including the traversal of references.
```
Comment (michaelb990) [inline release-criteria.md:49]: What is the "traversal of references"? This isn't talked about in our spec so far. I believe the only operations we're supporting do single hop lookups, so is this necessary?

File: release-criteria.md (hunk 1 above)
Comment (michaelb990) [reply release-criteria.md:62]: +1, I think this sentence is just missing some words. Can we rephrase it to be more clear?

**Review by michaelb990 — CHANGES_REQUESTED** [review]

File: release-criteria.md (hunk 2 above)
Comment (jonjohnsonjr) [reply release-criteria.md:17]: +1

`+json` is part of the [structured syntax suffix registry](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml).

**Comment by SteveLasker** [comment]
Closing in lieu of #35
